llvm/docs/HowToUpdateDebugInfo.rst

```diff
@@ -91,8 +91,12 @@ misattributed to a block containing one of the instructions-to-be-merged.
 
 Examples of transformations that should follow this rule include:
 
-* Merging identical loads/stores which occur on both sides of a CFG diamond
-  (see the ``MergedLoadStoreMotion`` pass).
+* Hoisting identical instructions from successors of a conditional branch or
```

Comment:
Is there a way to clarify that this must be from /all/ predecessors or successors (like if you sink something from 2 predecessors into a block that has 3 - you can't use a merged location there, because that makes the location from the first 2 predecessors "reachable" when the 3rd predecessor is taken)

Reply:
Actually, the merging updater (eg., `applyMergedLocation`) will drop the specific row or column number in the debug location when merging different `DebugLoc`s, only keeping their common scope. As a result, the merged debug location would not provide "extra coverage". Moreover, instructions of different predecessors (ie., at least two instructions) are sunk only when they are all identical (for now, I haven't seen any exception). So, I think it's not necessary.

Reply:
@dwblaikie Hi, I think I find the case you mentioned – the test [hoist-drop-debug-info.ll](https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/GVNHoist/hoist-drop-debug-info.ll) for GVNHoist. In this test, the control flow is shown below:
```
   entry
 /   |   \
sw1 sw2 exit1
 \   |
  exit
```

There is a store instruction in each of `sw1`, `sw2`, and `exit1`, and GVNHoist replace all the store with a merged store in `entry`. As a result, GVNHoist drops the debug location of the merged store. [Here](https://github.com/llvm/llvm-project/blob/d9b6e9f1c1565d9469eb0546da8c276051175408/llvm/lib/Transforms/Scalar/GVNHoist.cpp#L1118) is the pass code. (IMO, since all stores in all paths through `entry` are hoisted, the debug locations could be merged. However, in this case they are dropped. I think the reason would be that `exit1` does not go to `exit`?)

Does this conform to what you say?

Reply:
if every path (sw1, sw2, exit1) out of a block has the same location, then I think it's OK to merge and preserve, rather than drop. So I don't think this is the scenario I was describing above - where two out of three (sw1 and sw2, but not exit1) might share an expression, then it has to be dropped.

Reply:
> if every path (sw1, sw2, exit1) out of a block has the same location, then I think it's OK to merge and preserve, rather than drop.

So, could this be a mishandled update? (I have not found any case indicating that here needs a drop. Just as you said, they can be merged.)

> So I don't think this is the scenario I was describing above - where two out of three (sw1 and sw2, but not exit1) might share an expression, then it has to be dropped.

I understand whay you mean. However, I could't find such a case right now. Or maybe I can try to describe the scenario in natural language with a brief example (For example, adding a small code snippet below)?

Reply:
This paragraph already doesn't have concrete examples, so maybe it's just about adding the word "all" in the sentence "Hoisting identical instructions from /all/ successors of a conditional branch ..."

Reply:
@dwblaikie Thanks for your suggestion!